# Proyek Akhir: Menyelesaikan Permasalahan Perusahaan Edutech

- Nama: Alfiki Diastama Afan Firdaus
- Email: alfiki.diastama@gmail.com
- Id Dicoding: alfikiafan

## Persiapan

### Menyiapkan library yang dibutuhkan

In [2]:
import pandas as pd
import plotly.express as px
import plotly.figure_factory as ff
import plotly.graph_objects as go
import joblib
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score, roc_curve

### Menyiapkan data yang akan digunakan

In [194]:
df = pd.read_csv('data.csv', sep=';')

## Data Understanding

### Memeriksa informasi dasar dari dataset

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   Marital_status                                4424 non-null   int64  
 1   Application_mode                              4424 non-null   int64  
 2   Application_order                             4424 non-null   int64  
 3   Course                                        4424 non-null   int64  
 4   Daytime_evening_attendance                    4424 non-null   int64  
 5   Previous_qualification                        4424 non-null   int64  
 6   Previous_qualification_grade                  4424 non-null   float64
 7   Nacionality                                   4424 non-null   int64  
 8   Mothers_qualification                         4424 non-null   int64  
 9   Fathers_qualification                         4424 non-null   i

### Melihat lima baris pertama dari dataset

In [8]:
df.head()

Marital_status  Application_mode  Application_order  Course  \
0               1                17                  5     171   
1               1                15                  1    9254   
2               1                 1                  5    9070   
3               1                17                  2    9773   
4               2                39                  1    8014   

   Daytime_evening_attendance  Previous_qualification  \
0                           1                       1   
1                           1                       1   
2                           1                       1   
3                           1                       1   
4                           0                       1   

   Previous_qualification_grade  Nacionality  Mothers_qualification  \
0                         122.0            1                     19   
1                         160.0            1                      1   
2                         122.0            1                     37   
3                         122.0            1                     38   
4                         100.0            1                     37   

   Fathers_qualification  ...  Curricular_units_2nd_sem_credited  \
0                     12  ...                                  0   
1                      3  ...                                  0   
2                     37  ...                                  0   
3                     37  ...                                  0   
4                     38  ...                                  0   

   Curricular_units_2nd_sem_enrolled  Curricular_units_2nd_sem_evaluations  \
0                                  0                                     0   
1                                  6                                     6   
2                                  6                                     0   
3                                  6                                    10   
4                                  6                                     6   

   Curricular_units_2nd_sem_approved  Curricular_units_2nd_sem_grade  \
0                                  0                        0.000000   
1                                  6                       13.666667   
2                                  0                        0.000000   
3                                  5                       12.400000   
4                                  6                       13.000000   

   Curricular_units_2nd_sem_without_evaluations  Unemployment_rate  \
0                                             0               10.8   
1                                             0               13.9   
2                                             0               10.8   
3                                             0                9.4   
4                                             0               13.9   

   Inflation_rate   GDP    Status  
0             1.4  1.74   Dropout  
1            -0.3  0.79  Graduate  
2             1.4  1.74   Dropout  
3            -0.8 -3.12  Graduate  
4            -0.3  0.79  Graduate  

[5 rows x 37 columns]

### Melihat informasi statistik dari dataset

In [15]:
df.describe(include='all')

Marital_status  Application_mode  Application_order       Course  \
count      4424.000000       4424.000000        4424.000000  4424.000000   
unique             NaN               NaN                NaN          NaN   
top                NaN               NaN                NaN          NaN   
freq               NaN               NaN                NaN          NaN   
mean          1.178571         18.669078           1.727848  8856.642631   
std           0.605747         17.484682           1.313793  2063.566416   
min           1.000000          1.000000           0.000000    33.000000   
25%           1.000000          1.000000           1.000000  9085.000000   
50%           1.000000         17.000000           1.000000  9238.000000   
75%           1.000000         39.000000           2.000000  9556.000000   
max           6.000000         57.000000           9.000000  9991.000000   

        Daytime_evening_attendance  Previous_qualification  \
count                  4424.000000             4424.000000   
unique                         NaN                     NaN   
top                            NaN                     NaN   
freq                           NaN                     NaN   
mean                      0.890823                4.577758   
std                       0.311897               10.216592   
min                       0.000000                1.000000   
25%                       1.000000                1.000000   
50%                       1.000000                1.000000   
75%                       1.000000                1.000000   
max                       1.000000               43.000000   

        Previous_qualification_grade  Nacionality  Mothers_qualification  \
count                    4424.000000  4424.000000            4424.000000   
unique                           NaN          NaN                    NaN   
top                              NaN          NaN                    NaN   
freq                             NaN          NaN                    NaN   
mean                      132.613314     1.873192              19.561935   
std                        13.188332     6.914514              15.603186   
min                        95.000000     1.000000               1.000000   
25%                       125.000000     1.000000               2.000000   
50%                       133.100000     1.000000              19.000000   
75%                       140.000000     1.000000              37.000000   
max                       190.000000   109.000000              44.000000   

        Fathers_qualification  ...  Curricular_units_2nd_sem_credited  \
count             4424.000000  ...                        4424.000000   
unique                    NaN  ...                                NaN   
top                       NaN  ...                                NaN   
freq                      NaN  ...                                NaN   
mean                22.275316  ...                           0.541817   
std                 15.343108  ...                           1.918546   
min                  1.000000  ...                           0.000000   
25%                  3.000000  ...                           0.000000   
50%                 19.000000  ...                           0.000000   
75%                 37.000000  ...                           0.000000   
max                 44.000000  ...                          19.000000   

        Curricular_units_2nd_sem_enrolled  \
count                         4424.000000   
unique                                NaN   
top                                   NaN   
freq                                  NaN   
mean                             6.232143   
std                              2.195951   
min                              0.000000   
25%                              5.000000   
50%                              6.000000   
75%                              7.000000   
max                             23.000000   

        Curricular_

### Mengecek apakah ada data yang hilang

In [6]:
df.isnull().sum()

Marital_status                                  0
Application_mode                                0
Application_order                               0
Course                                          0
Daytime_evening_attendance                      0
Previous_qualification                          0
Previous_qualification_grade                    0
Nacionality                                     0
Mothers_qualification                           0
Fathers_qualification                           0
Mothers_occupation                              0
Fathers_occupation                              0
Admission_grade                                 0
Displaced                                       0
Educational_special_needs                       0
Debtor                                          0
Tuition_fees_up_to_date                         0
Gender                                          0
Scholarship_holder                              0
Age_at_enrollment                               0


### Mengecek apakah ada data yang terduplikasi

In [14]:
duplicates = df.duplicated().sum()
print(f"Jumlah data yang terduplikasi: {duplicates}")

Jumlah data yang terduplikasi: 0


**Hasil pemeriksaan data:**  
- Dataset terdiri atas 37 kolom dan 4424 baris
- Tipe data terdiri atas data numerik (int64, float64) dan data kategorikal (object)
- Terdapat data yang seharusnya bertipe kategorikal tetapi format asal masih numerik (int64)
- Dataset tidak memiliki data yang hilang
- Dataset tidak memiliki data yang terduplikasi


**Kolom yang akan diubah ke tipe data `object`**:
* `Marital_status`
* `Application_mode`
* `Course`
* `Daytime_evening_attendance`
* `Previous_qualification`
* `Nacionality`
* `Mothers_qualification`
* `Fathers_qualification`
* `Mothers_occupation`
* `Fathers_occupation`
* `Displaced`
* `Educational_special_needs`
* `Debtor`
* `Tuition_fees_up_to_date`
* `Gender`
* `Scholarship_holder`
* `International`

In [191]:
categorical_columns = [
    'Marital_status', 'Application_mode', 'Course', 'Daytime_evening_attendance', 
    'Previous_qualification', 'Nacionality', 'Mothers_qualification', 'Fathers_qualification', 
    'Mothers_occupation', 'Fathers_occupation', 'Displaced', 'Educational_special_needs', 
    'Debtor', 'Tuition_fees_up_to_date', 'Gender', 'Scholarship_holder', 'International', 'Status'
]

for col in categorical_columns:
    df[col] = df[col].astype('object')

df.dtypes

Marital_status                                   object
Application_mode                                 object
Application_order                                 int64
Course                                           object
Daytime_evening_attendance                       object
Previous_qualification                           object
Previous_qualification_grade                    float64
Nacionality                                      object
Mothers_qualification                            object
Fathers_qualification                            object
Mothers_occupation                               object
Fathers_occupation                               object
Admission_grade                                 float64
Displaced                                        object
Educational_special_needs                        object
Debtor                                           object
Tuition_fees_up_to_date                          object
Gender                                          

## Exploratory Data Analysis (EDA)

### Melihat distribusi target (status mahasiswa)

In [186]:
status_counts = df['Status'].value_counts()

fig = px.bar(status_counts,
             title='Distribusi Status Mahasiswa',
             labels={'Status': 'Status Mahasiswa', 'value': 'Jumlah'},
             color=status_counts.index,
             color_discrete_sequence=['#1f77b4', '#ff7f0e', '#2ca02c']
            )

fig.update_layout(
    bargap=0.2,
    plot_bgcolor='white',
    showlegend=False,
    xaxis=dict(showline=True, linewidth=2, linecolor='black'),
    yaxis=dict(showline=True, linewidth=2, linecolor='black', showgrid=True, gridcolor='lightgrey', gridwidth=0.5),
)
fig.show()


**Insight**:  
Jumlah mahasiswa Graduate lebih banyak daripada Dropout dan Enrolled. Ini menunjukkan bahwa sebagian besar mahasiswa berhasil menyelesaikan studi mereka, namun masalah dropout tetap perlu diperhatikan, karena jumlahnya lebih dari separuh mahasiswa yang graduate.

### Analisis Fitur Kategorikal vs Status

In [187]:
category_maps = {
    'Marital_status': {
        1: 'Belum Menikah',
        2: 'Menikah',
        3: 'Janda/Duda',
        4: 'Cerai',
        5: 'Union De Facto',
        6: 'Pisah Secara Hukum'
    },
    'Daytime_evening_attendance': {
        1: 'Siang',
        0: 'Malam'
    },
    'Debtor': {
        1: 'Memiliki Utang',
        0: 'Tidak Memiliki Utang'
    },
    'Tuition_fees_up_to_date': {
        1: 'Tertib Pembayaran',
        0: 'Tidak Tertib Pembayaran'
    },
    'Gender': {
        1: 'Pria',
        0: 'Wanita'
    },
    'Scholarship_holder': {
        1: 'Penerima Beasiswa',
        0: 'Bukan Penerima Beasiswa'
    },
    'International': {
        1: 'Mahasiswa Internasional',
        0: 'Mahasiswa Lokal'
    }
}

feature_titles = {
    'Marital_status': 'Status Perkawinan',
    'Daytime_evening_attendance': 'Waktu Kuliah (Siang/Malam)',
    'Debtor': 'Status Peminjaman',
    'Tuition_fees_up_to_date': 'Status Pembayaran Uang Kuliah',
    'Gender': 'Jenis Kelamin',
    'Scholarship_holder': 'Penerima Beasiswa',
    'International': 'Mahasiswa Internasional'
}

def map_categories(df, category_maps):
    df_mapped = df.copy()
    for feature, mapping in category_maps.items():
        if feature in df_mapped.columns:
            df_mapped[feature] = df_mapped[feature].map(mapping)
    return df_mapped

df_mapped = map_categories(df, category_maps)

cat_features = ['Marital_status', 'Daytime_evening_attendance', 'Debtor',
                'Tuition_fees_up_to_date', 'Gender', 'Scholarship_holder', 'International']

color_map = {'Graduate': '#1f77b4', 'Enrolled': '#2ca02c', 'Dropout': '#ff7f0e'}

for feature in cat_features:
    df_feature = df_mapped.groupby([feature, 'Status'], observed=False).size().reset_index(name='count')
    df_feature['percentage'] = df_feature.groupby(feature, observed=False)['count'].transform(lambda x: x / x.sum() * 100)
    
    title = f'Distribusi Status Berdasarkan {feature_titles[feature]}'
    
    fig = px.bar(df_feature, x=feature, y='percentage', color='Status',
                 barmode='group',
                 title=title,
                 color_discrete_map=color_map)
    
    fig.update_layout(
        bargap=0.2,
        plot_bgcolor='white',
        xaxis=dict(
            showline=True, 
            linewidth=2, 
            linecolor='black', 
            title='Kategori'
        ),
        yaxis=dict(
            showline=True, 
            linewidth=2, 
            linecolor='black', 
            showgrid=True, 
            gridcolor='lightgrey', 
            gridwidth=0.5, 
            title='Persentase'
        ),
    )

    fig.show()

**Insights:**  

1. **Status Perkawinan dan Dropout**:

   Mahasiswa yang tidak berstatus **menikah** memiliki persentase **dropout** yang lebih tinggi dibandingkan yang **belum menikah**.

2. **Waktu Kuliah (Siang/Malam) dan Dropout**:

   Mahasiswa yang kuliah di waktu **siang** lebih cenderung **lulus** dibandingkan yang kuliah di waktu **malam**.
   Mahasiswa malam cenderung memiliki tingkat **dropout** yang lebih tinggi, yang mungkin terkait dengan tanggung jawab lain di luar kuliah.

3. **Status Peminjaman dan Dropout**:

   Mahasiswa yang memiliki **utang** cenderung memiliki tingkat **dropout** yang lebih tinggi dibandingkan yang tidak memiliki utang.
   Masalah finansial bisa jadi menjadi salah satu alasan utama mahasiswa memilih untuk keluar dari program studi.

4. **Status Pembayaran Uang Kuliah dan Dropout**:

   Mahasiswa yang tidak **tertib dalam pembayaran uang kuliah** menunjukkan tingkat **dropout** yang sangat tinggi.
   Pembayaran yang tidak teratur berpotensi menjadi faktor penghambat bagi kelangsungan studi mahasiswa.

5. **Jenis Kelamin dan Dropout**:

   **Pria** memiliki persentase **dropout** yang lebih tinggi dibandingkan **wanita**.

6. **Penerima Beasiswa dan Dropout**:

   Mahasiswa yang **bukan penerima beasiswa** cenderung memiliki tingkat **dropout** yang lebih tinggi dibandingkan yang menerima beasiswa, yang bisa menunjukkan kalau dukungan finansial punya kontribusi agar mahasiswa tetap dapat belajar.

7. **Mahasiswa Internasional dan Dropout**:

   Mahasiswa **internasional** dan **domestik** memiliki tingkat **dropout** yang hampir sama. Mungkin perbedaan dalam kelompok ini tidak terlalu memengaruhi keputusan mahasiswa untuk keluar dari program studi.


### Analisis Fitur Numerik vs Status (Boxplot)

In [188]:
num_features = ['Admission_grade', 'Previous_qualification_grade',
                'Age_at_enrollment', 'Curricular_units_1st_sem_grade',
                'Curricular_units_2nd_sem_grade', 'Unemployment_rate', 'Inflation_rate', 'GDP']

color_map = {'Graduate': '#1f77b4', 'Enrolled': '#2ca02c', 'Dropout': '#ff7f0e'}

feature_titles = {
    'Admission_grade': 'Nilai Ujian Penerimaan Mahasiswa',
    'Previous_qualification_grade': 'Nilai Akhir Saat Pendidikan Sebelumnya',
    'Age_at_enrollment': 'Usia Saat Pendaftaran',
    'Curricular_units_1st_sem_grade': 'Nilai Mata Kuliah Semester 1',
    'Curricular_units_2nd_sem_grade': 'Nilai Mata Kuliah Semester 2',
    'Unemployment_rate': 'Tingkat Pengangguran',
    'Inflation_rate': 'Tingkat Inflasi',
    'GDP': 'PDB (Produk Domestik Bruto)'
}

for feature in num_features:
    fig = px.box(df, x='Status', y=feature, color='Status',
                 title=f'{feature_titles[feature]} Berdasarkan Status Mahasiswa',
                 color_discrete_map=color_map)
    
    fig.update_layout(
        bargap=0.2,
        plot_bgcolor='white',
        xaxis=dict(
            showline=True,
            linewidth=2,
            linecolor='black',
            title='Status Mahasiswa'
        ),
        yaxis=dict(
            showline=True,
            linewidth=2,
            linecolor='black',
            showgrid=True,
            gridcolor='lightgrey',
            gridwidth=0.5,
            title=feature_titles[feature]
        ),
    )

    fig.show()

**Insight:**  

1. **Nilai Ujian Penerimaan**  
   Mahasiswa yang **dropout** cenderung memiliki nilai ujian penerimaan yang lebih rendah dibandingkan mahasiswa yang **graduate** atau **enrolled**. Nilai ujian penerimaan yang rendah bisa menjadi indikator kuat masalah akademik yang berkontribusi pada keputusan untuk dropout.

2. **Nilai Akhir Pendidikan Sebelumnya**  
   Mahasiswa dengan nilai akhir pendidikan sebelumnya yang lebih rendah lebih cenderung mengalami **dropout**. Hal ini menunjukkan bahwa persiapan akademik yang kurang optimal sebelum masuk kuliah bisa menjadi faktor penting dalam keputusan untuk tidak melanjutkan studi.

3. **Usia Saat Pendaftaran**  
   Mahasiswa yang lebih tua pada saat pendaftaran cenderung memiliki tingkat **dropout** yang lebih tinggi. Hal ini mungkin terkait dengan tanggung jawab tambahan atau kurangnya kesiapan akademik yang lebih menantang.

4. **Nilai Mata Kuliah Semester 1 dan 2**  
   Mahasiswa dengan nilai yang lebih rendah di semester pertama dan kedua cenderung lebih sering **dropout**. Hal ini menunjukkan pentingnya kinerja akademis di awal perkuliahan sebagai prediktor keberhasilan studi.

5. **Tingkat Pengangguran, Inflasi, dan PDB**  
   Tingkat pengangguran, inflasi, dan PDB negara asal mahasiswa tidak terlalu berpengaruh terhadap keputusan untuk **dropout**. Hal ini menunjukkan bahwa faktor ekonomi makro mungkin bukan menjadi penentu utama dalam keputusan mahasiswa untuk melanjutkan pendidikan.

### Korelasi antar fitur numerik

In [193]:
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns
corr_matrix = df[numeric_cols].corr()

corr_matrix

Application_order  ...       GDP
Application_order                                      1.000000  ...  0.030201
Previous_qualification_grade                          -0.064484  ... -0.052620
Admission_grade                                       -0.096930  ... -0.019519
Age_at_enrollment                                     -0.271154  ... -0.064678
Curricular_units_1st_sem_credited                     -0.133354  ... -0.026513
Curricular_units_1st_sem_enrolled                     -0.016808  ... -0.026262
Curricular_units_1st_sem_evaluations                  -0.092156  ... -0.099761
Curricular_units_1st_sem_approved                      0.035580  ...  0.018459
Curricular_units_1st_sem_grade                         0.058308  ...  0.054801
Curricular_units_1st_sem_without_evaluations          -0.031699  ... -0.144673
Curricular_units_2nd_sem_credited                     -0.125815  ... -0.024491
Curricular_units_2nd_sem_enrolled                      0.028878  ... -0.007592
Curricular_units_2nd_sem_evaluations                  -0.055089  ... -0.004854
Curricular_units_2nd_sem_approved                      0.071793  ...  0.022427
Curricular_units_2nd_sem_grade                         0.055517  ...  0.071269
Curricular_units_2nd_sem_without_evaluations          -0.015757  ... -0.080292
Unemployment_rate                                     -0.098419  ... -0.335178
Inflation_rate                                        -0.011133  ... -0.112295
GDP                                                    0.030201  ...  1.000000

[19 rows x 19 columns]

In [192]:
numerics = df.select_dtypes(include='number')
corr = numerics.corr()

fig = px.imshow(corr,
                labels=dict(color="Korelasi"),
                title='Matriks Korelasi Fitur Numerik (terhadap Status Dropout)',
                x=corr.columns,
                y=corr.columns,
                color_continuous_scale='RdBu',
                zmin=-1, zmax=1)

fig.update_layout(
    plot_bgcolor='white',
    title_font=dict(size=18, family='Arial, sans-serif'),
    height=800,
    width=800,
    xaxis=dict(title='Fitur'),
    yaxis=dict(title='Fitur'),
    coloraxis_colorbar=dict(
        title="Korelasi",
        ticks="outside"
    )
)

fig.show()

**Insights:**  

1. **Korelasi tinggi antara Curricular Units Semester 1 dan Semester 2**:

   Fitur seperti **Curricular\_units\_1st\_sem\_credited**, **Curricular\_units\_1st\_sem\_enrolled**, **Curricular\_units\_1st\_sem\_approved**, dan seterusnya, menunjukkan korelasi yang sangat tinggi dengan fitur serupa di **semester 2**.
   Hal ini menunjukkan bahwa jika mahasiswa memiliki kinerja akademik yang baik di semester pertama, mereka cenderung juga melanjutkan dan berkinerja baik di semester kedua.

2. **Korelasi antara Nilai Admission dan Previous Qualification**:

   Terdapat korelasi positif yang moderat antara **Previous\_qualification\_grade** dan **Admission\_grade**, yang menunjukkan bahwa nilai pendidikan sebelumnya dapat mempengaruhi hasil ujian penerimaan mahasiswa.

3. **Pengaruh pada Unemployment dan GDP**:

   **Tingkat pengangguran (Unemployment\_rate)** memiliki korelasi negatif yang cukup kuat dengan **GDP**, yang mengindikasikan bahwa ketika PDB lebih rendah, tingkat pengangguran mungkin lebih tinggi. Ini bisa berpengaruh pada keputusan mahasiswa untuk tetap melanjutkan pendidikan atau tidak.

4. **Kurangnya Korelasi Antara Faktor Ekonomi dan Akademik**:

   Fitur-fitur ekonomi seperti **Inflation\_rate** dan **GDP** tidak menunjukkan korelasi kuat dengan faktor akademik seperti **Curricular\_units** atau **Admission\_grade**. Ini menunjukkan bahwa kondisi ekonomi makro tidak terlalu memengaruhi langsung keputusan akademik mahasiswa dalam konteks dataset ini.

5. **Usia dan Performa Akademik**:

   **Age\_at\_enrollment** memiliki korelasi negatif lemah dengan beberapa variabel akademik seperti **Curricular\_units\_1st\_sem\_credited** dan **Curricular\_units\_2nd\_sem\_credited**, yang bisa menunjukkan bahwa mahasiswa yang lebih tua mungkin memiliki keterlibatan akademik yang sedikit lebih rendah pada awalnya.

### Analisis Multivariat

**Interaksi Antara Admission_grade, Debtor, dan Status**  
Apakah siswa yang nilai masuknya rendah dan punya utang cenderung dropout?

In [195]:
df_mapped = df.copy()

color_map = {'Graduate': '#1f77b4', 'Enrolled': '#2ca02c', 'Dropout': '#ff7f0e'}

df_mapped['Debtor'] = df_mapped['Debtor'].map({0: 'Tidak Punya Utang', 1: 'Punya Utang'})

fig = px.scatter(
    df_mapped, 
    x='Admission_grade', 
    y='Curricular_units_1st_sem_grade',
    color='Status',
    color_discrete_map=color_map,
    symbol='Debtor',
    title='Nilai Penerimaan vs Nilai Mata Kuliah Semester 1',
    hover_data=['Debtor'],
    category_orders={'Status': ['Dropout', 'Graduate', 'Enrolled']}
)

fig.update_layout(
    xaxis=dict(
        title='Nilai Penerimaan',
        showline=True,
        linewidth=1,
        linecolor='black'
    ),
    yaxis=dict(
        title='Nilai Mata Kuliah Semester 1',
        showline=True,
        linewidth=1,
        linecolor='black',
        showgrid=True,
        gridcolor='lightgrey'
    ),
    coloraxis_colorbar=dict(
        title='Status Mahasiswa'
    ),
    title='Nilai Penerimaan vs Nilai Mata Kuliah Semester 1',
    plot_bgcolor='white',
)

fig.show()

**Insights:**  

1. **Mahasiswa yang Dropout cenderung memiliki nilai ujian penerimaan yang lebih rendah**:

   Mahasiswa dengan status **Dropout** (baik yang memiliki utang atau tidak) umumnya memiliki nilai ujian penerimaan yang lebih rendah dan nilai mata kuliah semester 1 yang juga rendah.

2. **Mahasiswa yang Terdaftar dan Lulus memiliki nilai ujian penerimaan yang lebih tinggi**:

   Mahasiswa dengan status **Graduate** atau **Enrolled** memiliki nilai ujian penerimaan yang lebih tinggi, terutama yang **Enrolled** (terdaftar) dan **Graduate** (lulus).

3. **Tidak ada perbedaan mencolok berdasarkan status peminjaman (debtors vs non-debtors)**:
   Terlepas dari apakah mahasiswa memiliki utang atau tidak, distribusi nilai ujian penerimaan dan nilai mata kuliah semester 1 tampaknya cukup serupa. Hal ini bisa menunjukkan bahwa status peminjaman tidak secara langsung mempengaruhi performa akademik mahasiswa, meskipun faktor ekonomi bisa memiliki peran lain dalam keputusan untuk dropout.

**Interaksi Antara Tuition_fees_up_to_date, Scholarship_holder, dan Status**  
Apakah siswa yang tidak membayar tepat waktu dan tidak mendapat beasiswa lebih rentan dropout?

In [196]:
df_mapped = df.copy()

df_mapped['Tuition_fees_up_to_date'] = df_mapped['Tuition_fees_up_to_date'].map({1: 'Tertib Pembayaran', 0: 'Tidak Tertib Pembayaran'})
df_mapped['Scholarship_holder'] = df_mapped['Scholarship_holder'].map({1: 'Penerima Beasiswa', 0: 'Bukan Penerima Beasiswa'})

color_map = {
    'Tertib Pembayaran': '#1f77b4',
    'Tidak Tertib Pembayaran': '#ff7f0e',
    'Penerima Beasiswa': '#1f77b4',
    'Bukan Penerima Beasiswa': '#2ca02c',
    'Graduate': '#1f77b4',
    'Enrolled': '#2ca02c',
    'Dropout': '#ff7f0e'
}

fig = px.sunburst(
    df_mapped,
    path=['Tuition_fees_up_to_date', 'Scholarship_holder', 'Status'],
    values=None,
    title='Distribusi Status Mahasiswa Berdasarkan Pembayaran Uang Kuliah dan Penerima Beasiswa',
    color='Status',
    color_discrete_map=color_map
)

fig.update_traces(
    textinfo='label+percent entry',
    textfont=dict(size=12),
    textfont_weight='bold',
    hovertemplate='%{label}<br>%{percentEntry:.2%}<extra></extra>'
)

fig.update_layout(
    width=800,
    height=800
)

fig.show()

**Insights:**  

1. **Mayoritas Mahasiswa Bukan Penerima Beasiswa**:

   Sebagian besar mahasiswa tidak menerima beasiswa (64%) dan mereka lebih cenderung **lulus** (Graduate) dengan persentase 31% dari total mahasiswa.
   Mahasiswa yang **bukan penerima beasiswa** dan **tertib dalam pembayaran uang kuliah** cenderung memiliki lebih banyak mahasiswa **lulus** dibandingkan yang **dropout** atau **terdaftar**.

2. **Mahasiswa yang Tertib Pembayaran Kuliah Lebih Cenderung Lulus**:

   **82%** mahasiswa menunjukkan pembayaran uang kuliah yang tertib, dan mayoritas dari mereka lulus (**Graduate**) sebanyak 19%.
   Ini menegaskan bahwa mahasiswa yang dapat memenuhi kewajiban pembayaran lebih terorganisir dan memiliki kemungkinan lebih besar untuk berhasil menyelesaikan studi mereka.

3. **Dropout pada Mahasiswa yang Tidak Tertib Membayar**:

   **Dropout** lebih banyak ditemukan di antara mahasiswa yang tidak tertib dalam pembayaran uang kuliah (10% dari mahasiswa yang tidak tertib). Ini menunjukkan hubungan antara masalah keuangan dan keputusan untuk keluar dari pendidikan.

4. **Penerima Beasiswa Memiliki Tingkat Dropout Lebih Rendah**:

   Mahasiswa yang menerima beasiswa cenderung memiliki tingkat **dropout** yang lebih rendah. Hal ini bisa menunjukkan bahwa dukungan finansial (beasiswa) membantu mengurangi tekanan keuangan, yang berpotensi meningkatkan peluang keberhasilan dalam pendidikan.

**Interaksi: Age_at_enrollment, Educational_special_needs, dan Status**  
Apakah usia lebih tua dan kebutuhan khusus memengaruhi potensi dropout?

In [197]:
df_mapped = df.copy()

df_mapped['Educational_special_needs'] = df_mapped['Educational_special_needs'].map({0: 'Tidak Memiliki Kebutuhan Khusus', 1: 'Memiliki Kebutuhan Khusus'})

color_map = {
    'Graduate': '#1f77b4',
    'Enrolled': '#2ca02c',
    'Dropout': '#ff7f0e'
}

fig = px.box(
    df_mapped,
    x='Educational_special_needs',
    y='Age_at_enrollment',
    color='Status',
    title='Usia Masuk Berdasarkan Status Mahasiswa dan Kebutuhan Khusus',
    color_discrete_map=color_map
)

fig.update_layout(
    bargap=0.2,
    plot_bgcolor='white',
    xaxis=dict(
        showline=True,
        linewidth=2,
        linecolor='black',
        title='Apakah Berkebutuhan Khusus?'
    ),
    yaxis=dict(
        showline=True,
        linewidth=2,
        linecolor='black',
        showgrid=True,
        gridcolor='lightgrey',
        gridwidth=0.5,
        title='Usia Saat Pendaftaran'
    ),
)

fig.show()

**Insights:**  

Mahasiswa yang lebih tua cenderung memiliki tingkat **dropout** yang lebih tinggi, terutama di antara mereka yang memiliki kebutuhan pendidikan khusus. Ini menunjukkan bahwa mahasiswa dengan kebutuhan pendidikan khusus mungkin menghadapi tantangan tambahan yang dapat memengaruhi keputusan mereka untuk melanjutkan studi.

## Data Preparation / Preprocessing

**Menghapus data dengan status "Enrolled"**

In [225]:
df_model = df[df['Status'].isin(['Graduate', 'Dropout'])].copy()

**Encode label dan fitur kategorikal**

In [226]:
df_model['Status'] = df_model['Status'].map({'Dropout': 1, 'Graduate': 0})

categorical_cols = df_model.select_dtypes(include='object').columns
le = LabelEncoder()
for col in categorical_cols:
    df_model[col] = le.fit_transform(df_model[col])

joblib.dump(le, 'model/label_encoder.joblib')

['model/label_encoder.joblib']

**Feature Engineering**  
Academic_consistency: selisih antara nilai semester 1 & 2

In [227]:
df_model['Academic_consistency'] = abs(df_model['Curricular_units_1st_sem_grade'] - df_model['Curricular_units_2nd_sem_grade'])

Total_failed_units: total mata kuliah yang tidak lulus

In [228]:
df_model['Total_failed_units'] = (df_model['Curricular_units_1st_sem_enrolled'] - df_model['Curricular_units_1st_sem_approved']) + \
    (df_model['Curricular_units_2nd_sem_enrolled'] - df_model['Curricular_units_2nd_sem_approved'])

Total_evaluations: total evaluasi semester 1 + 2

In [229]:
df_model['Total_evaluations'] = df_model['Curricular_units_1st_sem_evaluations'] + df_model['Curricular_units_2nd_sem_evaluations']

model_columns = df_model.drop('Status', axis=1).columns.tolist()
joblib.dump(model_columns, 'model/model_columns.joblib')

['model/model_columns.joblib']

**Memisahkan data menjadi data latih dan data uji**

In [230]:
X = df_model.drop('Status', axis=1)
y = df_model['Status']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

**Feature Scaling**

In [231]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

joblib.dump(scaler, 'model/scaler.joblib')

['model/scaler.joblib']

## Modeling

### Logistic Regression

In [232]:
logreg = LogisticRegression(solver='saga', max_iter=10000, random_state=42)
logreg.fit(X_train, y_train)

joblib.dump(logreg, 'model/logreg_model.joblib')

['model/logreg_model.joblib']

### Random Forest

In [233]:
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

joblib.dump(rf, 'model/rf_model.joblib')

['model/rf_model.joblib']

## Evaluation

**Classification Report**

In [234]:
y_pred_log = logreg.predict(X_test)
y_prob_log = logreg.predict_proba(X_test)[:, 1]
print("Logistic Regression Report:\n", classification_report(y_test, y_pred_log))
print("ROC AUC (Logistic Regression):", roc_auc_score(y_test, y_prob_log))

y_pred_rf = rf.predict(X_test)
y_prob_rf = rf.predict_proba(X_test)[:, 1]
print("\nRandom Forest Report:\n", classification_report(y_test, y_pred_rf))
print("ROC AUC (Random Forest):", roc_auc_score(y_test, y_prob_rf))

Logistic Regression Report:
               precision    recall  f1-score   support

           0       0.87      0.98      0.92       442
           1       0.96      0.77      0.86       284

    accuracy                           0.90       726
   macro avg       0.91      0.88      0.89       726
weighted avg       0.90      0.90      0.90       726

ROC AUC (Logistic Regression): 0.9444108087438658

Random Forest Report:
               precision    recall  f1-score   support

           0       0.94      0.95      0.94       442
           1       0.92      0.90      0.91       284

    accuracy                           0.93       726
   macro avg       0.93      0.92      0.93       726
weighted avg       0.93      0.93      0.93       726

ROC AUC (Random Forest): 0.9741770760308457


**Confusion Matrix**

In [235]:
conf_matrix_logreg = confusion_matrix(y_test, y_pred_log)
fig_logreg = ff.create_annotated_heatmap(
    z=conf_matrix_logreg,
    x=['Predicted Graduate', 'Predicted Dropout'],
    y=['Actual Graduate', 'Actual Dropout'],
    colorscale='Blues',
    showscale=True
)
fig_logreg.update_layout(
    title="Confusion Matrix - Logistic Regression",
    width=500,
    height=500
)
fig_logreg.show()

conf_matrix_rf = confusion_matrix(y_test, y_pred_rf)
fig_rf = ff.create_annotated_heatmap(
    z=conf_matrix_rf,
    x=['Predicted Graduate', 'Predicted Dropout'],
    y=['Actual Graduate', 'Actual Dropout'],
    colorscale='Blues',
    showscale=True
)
fig_rf.update_layout(
    title="Confusion Matrix - Random Forest",
    width=500,
    height=500
)
fig_rf.show()

In [224]:
fpr_log, tpr_log, _ = roc_curve(y_test, y_prob_log)
fpr_rf, tpr_rf, _ = roc_curve(y_test, y_prob_rf)

fig = go.Figure()
fig.add_trace(go.Scatter(x=fpr_log, y=tpr_log, name='Logistic Regression', line=dict(width=2)))
fig.add_trace(go.Scatter(x=fpr_rf, y=tpr_rf, name='Random Forest', line=dict(width=2)))
fig.add_trace(go.Scatter(x=[0, 1], y=[0, 1], name='Baseline', line=dict(dash='dash')))

fig.update_layout(title='ROC Curve Comparison',
                  xaxis_title='False Positive Rate',
                  yaxis_title='True Positive Rate',
                  plot_bgcolor='white',
                  width=650,
                  height=500)
fig.show()